polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -328,3 +215,176 @@ fn runtime_construction_check(
+					let memory_tracker_stats =
+						get_memory_tracker_loop_stats(memory_tracker_thread, process::id()).await;
+
+					let memory_stats = MemoryStats {
+						#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
+						memory_tracker_stats,
+						#[cfg(target_os = "linux")]
+						max_rss: extract_max_rss_stat(max_rss, process::id()),
+					};
+
+					let response: Result<Response, PrepareError> =
+						Ok(Response { artifact, memory_stats });
+
+					pipe_write
+						.write_all(response.encode().as_slice())
+						.unwrap_or_else(|_| process::exit(libc::EXIT_FAILURE));
+				},
+				Err(err) => {
+					pipe_write
+						.write_all(Err::<Response, PrepareError>(err).encode().as_slice())
+						.unwrap_or_else(|_| process::exit(libc::EXIT_FAILURE));
+				},
+			}
+
+			process::exit(libc::EXIT_SUCCESS);
+		},
+		_ => process::exit(libc::EXIT_FAILURE),
+	}
+}
+
+async fn handle_parent_process(
+	mut pipe_read: os_pipe::PipeReader,
+	temp_artifact_dest: PathBuf,
+	worker_pid: u32,
+	usage_before: Usage,
+	timeout: u64,
+) -> Result<PrepareStats, PrepareError> {
+	let mut received_data = Vec::new();
+
+	pipe_read
+		.read_to_end(&mut received_data)
+		.map_err(|err| PrepareError::Panic(err.to_string()))?;
+	let status = nix::sys::wait::wait();
+	let usage_after = nix::sys::resource::getrusage(UsageWho::RUSAGE_CHILDREN)
+		.map_err(|err| PrepareError::Panic(err.to_string()))?;
+	let cpu_tv = (get_total_cpu_usage(usage_after) - get_total_cpu_usage(usage_before)) as u64;
+
+	return match status {
+		Ok(nix::sys::wait::WaitStatus::Exited(_, libc::EXIT_SUCCESS)) => {
+			let result: Result<Response, PrepareError> = parity_scale_codec::decode_from_bytes(
+				bytes::Bytes::copy_from_slice(received_data.as_slice()),
+			)
+			.map_err(|e| PrepareError::Panic(e.to_string()))?;
+			match result {
+				Err(err) => Err(err),
+				Ok(response) => {
+					// Write the serialized artifact into a temp file.
+					//
+					// PVF host only keeps artifacts statuses in its memory,
+					// successfully compiled code gets stored on the disk (and
+					// consequently deserialized by execute-workers). The prepare worker
+					// is only required to send `Ok` to the pool to indicate the
+					// success.
+					gum::debug!(
+					target: LOG_TARGET,
+					%worker_pid,
+					"worker: writing artifact to {}",
+					temp_artifact_dest.display(),
+					);
+					if let Err(err) =
+						tokio::fs::write(&temp_artifact_dest, &response.artifact).await
+					{
+						return Err(PrepareError::Panic(format!("{:?}", err)))
+					};
+
+					Ok(PrepareStats {
+						memory_stats: response.memory_stats,
+						cpu_time_elapsed: Duration::from_secs(cpu_tv as u64),
+					})
+				},
+			}
+		},
+		_ => {
+			if cpu_tv >= timeout {
+				return Err(PrepareError::TimedOut)
+			}
+			Err(PrepareError::Panic("child finished with unknown status".to_string()))
+		},
+	}
+}
+
+fn get_total_cpu_usage(rusage: Usage) -> u64 {
```

Comment:
1. This should have a docstring
2. It should return a `Duration`, otherwise the `u64` is ambiguous (is it the number of seconds? milliseconds? microseconds?)
3. We're returning the number of seconds results in a loss of resolution. We might as well do `Duration::from_micros` here for maximum precision.